src/codegen/llvm/codegen_llvm_visitor.cpp

```diff
@@ -468,31 +468,73 @@ void CodegenLLVMVisitor::visit_boolean(const ast::Boolean& node) {
     ir_builder.create_boolean_constant(node.get_value());
 }
 
-/**
- * Currently, this functions is very similar to visiting the binary operator. However, the
- * difference here is that the writes to the LHS variable must be atomic. These has a particular
- * use case in synapse kernels. For simplicity, we choose not to support atomic writes at this
- * stage and emit a warning.
- *
- * \todo support this properly.
- */
 void CodegenLLVMVisitor::visit_codegen_atomic_statement(const ast::CodegenAtomicStatement& node) {
-    if (platform.is_cpu_with_simd())
-        logger->warn("Atomic operations are not supported");
+    // Get the variable node that need an atomic update.
+    const auto& var = std::dynamic_pointer_cast<ast::VarName>(node.get_lhs());
+    if (!var)
+        throw std::runtime_error("Error: only 'VarName' update is supported\n");
 
-    // Support only assignment for now.
+    // Evaluate RHS of the update.
     llvm::Value* rhs = accept_and_get(node.get_rhs());
-    if (node.get_atomic_op().get_value() != ast::BinaryOp::BOP_ASSIGN)
-        throw std::runtime_error(
-            "Error: only assignment is supported for CodegenAtomicStatement\n");
-    const auto& var = dynamic_cast<ast::VarName*>(node.get_lhs().get());
-    if (!var)
-        throw std::runtime_error("Error: only 'VarName' assignment is supported\n");
 
-    // Process the assignment as if it was non-atomic.
-    if (platform.is_cpu_with_simd())
-        logger->warn("Treating write as non-atomic");
-    write_to_variable(*var, rhs);
+    // First, check if it is an atomic write only and we can return early.
+    // Otherwise, extract what kind of atomic update we want to make.
+    ast::BinaryOp atomic_op = node.get_atomic_op().get_value();
+    if (atomic_op == ast::BinaryOp::BOP_ASSIGN) {
+        write_to_variable(*var, rhs);
+        return;
+    }
+    ast::BinaryOp op = ir_builder.extract_atomic_op(atomic_op);
+
+    // For different platforms, we handle atomic updates differently!
+    if (platform.is_cpu_with_simd()) {
+        throw std::runtime_error("Error: no atomic update support for SIMD CPUs\n");
+    } else if (platform.is_gpu()) {
+        const auto& identifier = var->get_name();
+
+        // We only need to support atomic updates to instance struct members.
+        if (!identifier->is_codegen_instance_var())
+            throw std::runtime_error("Error: atomic updates for non-instance variable\n");
+
+        const auto& node = std::dynamic_pointer_cast<ast::CodegenInstanceVar>(identifier);
+        const auto& instance_name = node->get_instance_var()->get_node_name();
+        const auto& member_node = node->get_member_var();
+        const auto& member_name = member_node->get_node_name();
+
+        if (!instance_var_helper.is_an_instance_variable(member_name))
+            throw std::runtime_error("Error: " + member_name +
+                                     " is not a member of the instance variable\n");
```

Comment:
to clarify: is this somehow overlaps with line 496+497?

Reply:
After merge with `llvm` branch those lines are `539`, `540`

Reply:
I think that
```
if (!identifier->is_codegen_instance_var())
            throw std::runtime_error("Error: atomic updates for non-instance variable\n");
```
checks whether `identifier` is type of `InstanceStruct` and 
```
if (!instance_var_helper.is_an_instance_variable(member_name))
            throw std::runtime_error("Error: " + member_name +
                                     " is not a member of the instance variable\n");
```
if `member_name` is a variable defined in the mod file in such way that there will be multiple instance of it so it will be a pointer variable inside the `InstanceStruct`, for example a `RANGE` variable. The second check is because it could only be such variables that need atomic insturctions

Reply:
What @iomaganaris said is right, I think. First check is for `InstanceStruct`, second is that we are generating code for pointer variable (member of instance). I think that it will always be like this (based on helper visitor) - so it;s nore of a sanity check :)